src/blueprints/tv2/timeline-object-factories/tv2-atem-video-mixer-timeline-object-factory.ts

```diff
@@ -62,36 +62,71 @@ export class Tv2AtemVideoMixerTimelineObjectFactory implements Tv2VideoMixerTime
     return Tv2AtemLayer.DOWNSTREAM_KEYER
   }
 
-  public createProgramTimelineObject(id: string, sourceInput: number, enable: TimelineEnable): AtemMeTimelineObject {
+  public createUpstreamKeyerTimelineObject(downstreamKeyer: Tv2DownstreamKeyer, enable: TimelineEnable): AtemMeTimelineObject {
+    const downstreamKeyerNumber: number = downstreamKeyer.Number + 1
     return {
-      id: `${ID_PREFIX}${id}`,
+      id: `upstreamKeyer${downstreamKeyerNumber}`,
       enable,
       priority: 1,
-      layer: Tv2AtemLayer.PROGRAM,
+      layer: Tv2AtemLayer.CLEAN_UPSTREAM_KEYER,
       content: {
         deviceType: DeviceType.ATEM,
         type: AtemType.ME,
         me: {
-          input: sourceInput,
-          transition: AtemTransition.CUT
+          upstreamKeyers: [
+            {
+              upstreamKeyerId: downstreamKeyer.Number,
+              onAir: true,
+              mixEffectKeyType: 0,
+              flyEnabled: false,
+              fillSource: downstreamKeyer.Fill,
+              cutSource: downstreamKeyer.Key,
+              maskEnabled: false,
+              lumaSettings: {
+                clip: this.convertPercentageToAtemPercentageValue(downstreamKeyer.Clip),
+                gain: this.convertPercentageToAtemPercentageValue(downstreamKeyer.Gain),
+              }
+            }
+          ]
         }
       }
     }
   }
 
-  public createCleanFeedTimelineObject(id: string, sourceInput: number, enable: TimelineEnable): AtemMeTimelineObject {
+  public createProgramTimelineObject(id: string, sourceInput: number, enable: TimelineEnable, transition: number = AtemTransition.CUT, transitionSettings?: AtemTransitionSettings): AtemMeTimelineObject {
+    return this.createAtemMeTimelineObjectForLayer(
+      `${ID_PREFIX}${id}`,
+      Tv2AtemLayer.PROGRAM,
+      enable,
+      {
+        input: sourceInput,
+        transition,
+        transitionSettings
+      })
+  }
+
+  public createCleanFeedTimelineObject(id: string, sourceInput: number, enable: TimelineEnable, transition: number = AtemTransition.CUT, transitionSettings?: AtemTransitionSettings): AtemMeTimelineObject {
+    return this.createAtemMeTimelineObjectForLayer(
+      `${ID_PREFIX}${id}`,
+      Tv2AtemLayer.CLEAN_FEED,
+      enable, 
+      {
+        input: sourceInput,
+        transition,
+        transitionSettings
+      })
+  }
+
+  private createAtemMeTimelineObjectForLayer(id: string, layer: Tv2AtemLayer, enable: TimelineEnable, me: AtemMeTimelineObject['content']['me']): AtemMeTimelineObject {
```

Comment:
As this currently creates a timeline object for both `Tv2AtemLayer.PROGRAM` and `Tv2AtemLayer.CLEAN_FEED`, it would still take in the id as an argument. It is the method that calls this, that no longer would take in an id.

Reply:
You still don't gain much from this method